---
title: "presentation_data_bact"
output: html_document
date: "2024-09-11"
editor_options: 
  chunk_output_type: console
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(
  comment = '', fig.width = 10)
```

# Setup

```{r echo = T, results = 'hide'}
## ---- IMPORT PACKAGES ----
library(readxl)
library(tidyverse)
library(iNEXT)
library(GGally)
library(sbm)
library(FactoMineR)
library(vegan)
library(alluvial) #aluvial plot
library(MuMIn) # model selection
library(aricode) # NMI

library(factoextra)
library(corrplot)
library(RColorBrewer)

## ---- usefull function ----
main_theme = theme_minimal()+
  theme(line = element_blank(), 
        axis.line = element_line(colour = "black"),
        panel.border = element_blank(),
        axis.ticks =  element_line(colour = "black"),
        axis.text.x = element_text(colour = "black", size=22, face="italic", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(colour = "black", size=22, face="italic"),
        legend.title = element_text(colour = "black", size=20,
                                    hjust =0.5),

        legend.text = element_text(colour = "black", size=18),
        axis.title= element_text(size=28),
        strip.text = element_text(colour = "black", size=15, face ="italic"))

zscore_normalization <- function(data) {
  # Step 1: Rank the data
  ranked_data <- rank(data)
  n <- length(data)
  
  # Step 2: Calculate quantile positions
  quantile_positions <- (ranked_data) / (n + 1)
  
  # Step 3: Transform quantiles to z-scores
  mean_std_normal <- qnorm(quantile_positions)
  
  return(mean_std_normal)
}
## ---- IMPORT DATA ----

read_tsv("data/metadata.txt") %>%
  rename_with(~gsub(".", "_", ., fixed = TRUE)) -> metadata
read_tsv("data/sample_coordinates.tsv")-> coord
read_tsv("data/divesities_bact.tsv")%>%
  dplyr::select(starts_with("Hill_"),soil_code)-> div_bact

read_xlsx("data/Historique_Parcelles-Oracle.xlsx")%>%
  dplyr::select(`N° Parcelle`, ...14, ...17,...20,) %>%
  rename(soil_code = "N° Parcelle", cultu_2014 = "...14", cultu_2015 = "...17",cultu_2016 = "...20")%>%
  mutate(soil_code = paste0("BU_",str_replace(soil_code, "([A-Z]+)(\\d+)", "\\1_\\2"))) -> preced_cultu

read_xlsx("data/Soil_Parameters_oracle.xlsx")%>%
  dplyr::select(`Var niébé 2017`, `Var Sorgho 2017`)%>%
  rename(varnieb = "Var niébé 2017", varsorgh = "Var Sorgho 2017")-> variety

metadata%>%
  dplyr::select(pracul,modsem,FO_Qte ,denpoq, freqsar, NPK_qte, uree_qte)%>%
  mutate(pracul = as.factor(pracul),
                 modsem = as.factor(modsem))->agri_practice

agri_practice%>%  
  mutate(across(where(is.numeric),~decostand(.x,"standardize")[,1]))->agri_practice_standa

read_xlsx("data/Soil_Parameters_oracle.xlsx")%>%
  dplyr::select(`C (g kg-1)`,`C/N`, `CEC (cmolc/kg)`, `Ca (cmol/kg)`,`Mg (cmolc/kg)`, `K(cmolc/kg)`, `SBE (cmolc/kg)`,           
    `TS (%)`, Site...3, `N° site...4`)%>%
  rename(Site = "Site...3",N_site = "N° site...4", C = "C (g kg-1)",CN = "C/N", CEC = "CEC (cmolc/kg)", Ca = "Ca (cmol/kg)",Mg = "Mg (cmolc/kg)",K = "K(cmolc/kg)",SBE = "SBE (cmolc/kg)",           
         ,TS = "TS (%)" )%>%
  mutate( N_site = str_replace(as.character(N_site), "^[1-9]$", "0\\0"),
          soil_code = str_c(Site,N_site))%>%
  mutate(soil_code = str_c("BU_",Site,"_",N_site))%>%
  left_join(metadata%>%dplyr::select(pHeau,Pto,Kto,Nto,MO,soil_code), by = "soil_code")-> soil
```
# BACTERIA
## setup bacteria table
```{r echo = T, results = 'hide'}
bact_taxonomic_levels <- c("kingdom", "phylum", "class", "order", "family", 
                           "genus", "species")
#{{import data}}
read_tsv("data/Oracle_soils_16S_341F_785R_87_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2") %>%
  #{{format column names}}
  rename_with(~sub("\\_.*", "", .)) %>%
  rename_with(~gsub("-", "_", .)) %>%
  rename_with(~gsub("OR", "BU",.)) %>%
  rename(CONT_BU_ext = BU_T_extr) %>%
  rename(CONT_BU_PCR1 = T_1) %>%
  rename(CONT_BU_PCR2 = T_2) %>%
  #{{select samples from ORACLE project}}
  select(OTU, taxonomy, starts_with("CONT", ignore.case = FALSE), 
         starts_with("BU", ignore.case = FALSE)) %>%
  #{{keep only resequenced samples when available}}
  select(-BU_SER_10, -BU_ST_10, -BU_YL_19, -BU_TS_3) %>%
  #{{homogeneize resequenced sample names}}
  rename(BU_SER_10 = BU_SER_10_S) %>%
  rename(BU_ST_10 = BU_ST_10_S) %>%
  rename(BU_YL_19 = BU_YL_19_S) %>%
  rename(BU_TS_3 = BU_TS_3_S) %>%
  rename_with(~sub("\\_1$", "_01", .)) %>%
  rename_with(~sub("\\_2$", "_02", .)) %>%
  rename_with(~sub("\\_3$", "_03", .)) %>%
  rename_with(~sub("\\_4$", "_04", .)) %>%
  rename_with(~sub("\\_5$", "_05", .)) %>%
  rename_with(~sub("\\_6$", "_06", .)) %>%
  rename_with(~sub("\\_7$", "_07", .)) %>%
  rename_with(~sub("\\_8$", "_08", .)) %>%
  rename_with(~sub("\\_9$", "_09", .)) %>%
  #{{sequences with "No_hit" has to be deleted}}
  filter(grepl("Bacteria|Archaea",taxonomy)) %>%
  separate(taxonomy, bact_taxonomic_levels, sep = "[|]", extra = "drop") %>%
  #{{sequences affiliated to chloroplast and mitonchondria are excluded}}
  filter(order != "Chloroplast", family != "Mitochondria")%>%
  droplevels() -> bact_ASV


bact_tax <- bact_ASV %>%
  #{{merge taxonomy}}
  left_join(select(read_tsv("data/Oracle_soils_16S_341F_785R_87_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2"), OTU, taxonomy), by = "OTU") %>%
  #{{select only OTU number and taxonomy}}
  select(OTU,taxonomy) %>%  
  separate(taxonomy, bact_taxonomic_levels, sep = "[|]", extra = "drop") %>%
  #{{replace missing taxonomic information and homogeneize the terms}}
  mutate_if(is.character, ~sub("uncultured", "unidentified", .)) %>% 
  mutate_if(is.character, ~sub("uncultured_bacterium", "unidentified", .)) %>%
  mutate_if(is.character, ~sub("unidentified_bacterium", "unidentified", .)) %>%
  mutate_if(is.character, ~sub("[*]", "unidentified", .)) %>%
  mutate_if(is.character , replace_na, replace = "unidentified") %>%
  #{{replace unidentified by the higher taxonomic rank}}
  mutate(phylum = if_else(str_detect(phylum, "unidentified"), 
                          str_c("unidentified_", kingdom), phylum)) %>%
  mutate(class = if_else(str_detect(class, "unidentified"), 
                         str_c("unidentified_", phylum), class)) %>%
  mutate(order = if_else(str_detect(order, "unidentified"), 
                         str_c("unidentified_", class), order)) %>%
  mutate(family = if_else(str_detect(family, "unidentified"), 
                          str_c("unidentified_", order), family)) %>%
  mutate(genus = if_else(str_detect(genus, "unidentified"), 
                         str_c("unidentified_", family), genus)) %>%
  mutate(species = if_else(str_detect(species, "unidentified"), 
                           str_c("unidentified_", genus), species)) %>%
  mutate(phylum = if_else(str_detect(phylum, "metagenome"), 
                          str_c("unidentified_", kingdom), phylum)) %>%
  mutate(class = if_else(str_detect(class, "metagenome"), 
                         str_c("unidentified_", phylum), class)) %>%
  mutate(order = if_else(str_detect(order, "metagenome"), 
                         str_c("unidentified_", class), order)) %>%
  mutate(family = if_else(str_detect(family, "metagenome"), 
                          str_c("unidentified_", order), family)) %>%
  mutate(genus = if_else(str_detect(genus, "metagenome"), 
                         str_c("unidentified_", family), genus)) %>%
  mutate(species = if_else(str_detect(species, "metagenome"),
                           str_c("unidentified_", genus), species)) %>%
  #{{homogenize unidentified terms for taxonomic rank}}
  mutate(class = str_replace(class,
                             "unidentified_unidentified_",
                             "unidentified_")) %>%
  mutate(order = str_replace(order,
                             "unidentified_unidentified_unidentified_",
                             "unidentified_")) %>%
  mutate(order = str_replace(order,
                             "unidentified_unidentified_",
                             "unidentified_")) %>%
  mutate(family = str_replace(family,
                              "unidentified_unidentified_unidentified_unidentified_",
                              "unidentified_")) %>% 
  mutate(family = str_replace(family,
                              "unidentified_unidentified_unidentified_",
                              "unidentified_")) %>% 
  mutate(family = str_replace(family,
                              "unidentified_unidentified_",
                              "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_unidentified_unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_unidentified_unidentified_unidentified_",
                               "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_unidentified_unidentified_",
                               "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_unidentified_",
                               "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_",
                               "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_",
                               "unidentified_")) %>% 
  #{{homogenize terms for unidentified or uncharacterized species (sp)}}
  mutate(species = if_else(str_detect(species, "_sp"), 
                           str_c("unidentified_", genus), species)) %>% 
  mutate(species = str_replace(species, "unidentified_unidentified_",
                               "unidentified_")) %>% 
  #{{convert OTU column in character from numeric}}
  mutate_if(is.numeric, as.character)

#{{Extract control samples}}
d <- replace(bact_ASV, bact_ASV == 0, NA) %>%
  select(OTU, starts_with("CONT")) %>%
  gather("samples", "reads", -OTU) %>%
  filter(!is.na(reads))

#{{Extract samples}}
bact_ASV_decont <- replace(bact_ASV, bact_ASV == 0, NA) %>%
  pivot_longer(starts_with("BU"), names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads)) %>%
  #{{merge with control samples}}
  left_join(count(d, OTU, wt = reads), by = "OTU") %>%
  #{{substract abundance of control samples}}
  mutate(reads = case_when(
    is.na(n)  ~ reads,
    n > reads ~ 0,
    TRUE      ~ reads - n)) %>%
  select(-n) %>%
  spread(samples, reads, fill = 0) %>%
  select(-starts_with("CONT"))
```

## Fist approach : select only known species

```{r echo = T, results = 'hide'}
bact_tax%>%
  filter(!str_detect(species, "unidentified"))-> bact_tax_known

bact_tax_known[duplicated(bact_tax_known$species),]
bact_ASV_decont%>%
  filter(as.character(OTU) %in% bact_tax_known$OTU)-> bact_ASV_decont_known

bact_ASV_decont_known%>%
  pivot_longer(starts_with("BU"), names_to = "soil_code", values_to = "read")%>%
  group_by(species, soil_code)%>%
  summarise(read = sum(read))%>%
  pivot_wider(names_from = "soil_code", values_from = "read") -> OTU_bact_known


```

```{r}
length(unique(bact_ASV_decont_known$family))
length(unique(bact_ASV_decont$family))
```


```{r echo = T}

OTU_bact_known_t <- OTU_bact_known%>%as.data.frame%>%
  column_to_rownames(var = "species") %>%
  #{{transpose data}}
  t()


set.seed(2)
bact_OTUs_rarefied_t <-  OTU_bact_known_t %>% 
  rrarefy (min(rowSums(OTU_bact_known_t))) %>% 
  #{{create a new column with sum of reads for each OTU in all samples}}
  t() %>% 
  as_tibble(rownames = NA) %>%
  rownames_to_column(var = "OTU") %>%
  #{{delete all sample with = 0 in rarefied samples}}
  mutate(total = rowSums(across(where(is.numeric)))) %>% 
  filter(total != 0) %>%
  select(-total) %>%
  column_to_rownames(var = "OTU") %>% 
  t() %>%
  as_tibble(rownames = NA)

hist(bact_OTUs_rarefied_t%>%as.matrix%>%log1p)
```


### Clustering bact known

#### Unarrange matrix
```{r}

bact_OTUs_rarefied_t%>%
  as.matrix()%>%
  log1p%>%
  plotMyMatrix(dimLabels = c("soil_code", "Bact Known"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))
```

#### arrange matrix

```{r, results='hide'}
bact_OTUs_rarefied_t

if(file.exists("outputs/sbm_bact_known.Rdata")){
  load(file = "outputs/sbm_bact_known.Rdata")
}else{
  bact_OTUs_rarefied_t%>%
  as.matrix%>%
  log1p%>%
  estimateBipartiteSBM(
    model = 'poisson', 
    dimLabels = c("soil_code", "Bact_Known"),
    estimOptions = list(nbCores = 8,plot = F)) -> sbm_bact_known
  
  save(sbm_bact_known, file = "outputs/sbm_bact_known.Rdata")
  
}


sbm_bact_known$storedModels %>% knitr::kable()

data.frame(Clust_bact_known_sbm = sbm_bact_known$memberships$soil_code, 
                         soil_code = rownames(bact_OTUs_rarefied_t)) -> cluster_df
```

```{r}
sbm_bact_known$storedModels %>%  ggplot() + aes(x = nbBlocks, y = ICL)  + geom_line() + geom_point(alpha = 0.5)
```

```{r}
make_id_order_by_group_df <- function(.data, group_name){
.data%>%as.data.frame()%>%
  rownames_to_column("id")%>%
  mutate(id = as.numeric(id))%>%
  arrange({{group_name}})
} 
bact_id = make_id_order_by_group_df(sbm_bact_known$memberships[2], Bact_Known)


soil_id =  make_id_order_by_group_df(sbm_bact_known$memberships[1], soil_code)

bact_OTUs_rarefied_t[soil_id$id,bact_id$id]%>%
  as.matrix()%>%
    log1p%>%
  plotMyMatrix(dimLabels = c("soil_code", "Bact known"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(bact_id$Bact_Known))[-1]-0.5, col ="red", linetype = 5)+
  geom_hline(yintercept = abs(which(!duplicated(soil_id$soil_code))[-1]-nrow(soil_id)-1)+0.5, col = "red")+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))


```
```{r}
plot(sbm_bact_known, type = "expected")
```

```{r}
plot(sbm_bact_known, type  = 'meso', 
     plotOptions = list(edge.width = 0.8, vertex.size = c(1,0.1)))
```


# Clustering covariate

## Cultural History

```{r echo = T, results = 'hide'}
preced_cultu%>%
  filter(!(is.na(cultu_2014) | is.na(cultu_2015) | is.na(cultu_2016))) -> preced_cultu
preced_cultu%>%
  mutate(across(starts_with("cultu"), ~case_when(
    .x == "Ass_Ses+N" ~ "c+l", # c : cereal ; l = legumes
    .x == "Ass_M+N" ~ "c+l",
    .x == "Ass_S+N" ~ "c+l",
    .x == "Ass_S+M+N" ~ "c+l",
    .x == "Rot_Avec_Leg" ~ "l",
    .x == "Rot_Sans_Leg" ~ "c",
    .x == "Jachère" ~ "jachere"
  ))) -> preced_cultu_cere_legu

### disjonctif table usable by sbm
tab.disjonctif(preced_cultu_cere_legu$cultu_2014) -> cult2014_bis
tab.disjonctif(preced_cultu_cere_legu$cultu_2015) -> cult2015_bis
tab.disjonctif(preced_cultu_cere_legu$cultu_2016) -> cult2016_bis

### check if colnames concord with eachother
colnames(cult2014_bis)
colnames(cult2015_bis)
colnames(cult2016_bis)

cult2014_bis = cbind(cult2014_bis,jachere = 0)
```

### Plot cultural history 
```{r echo = T}

cult2014_bis%>%
  t()%>%
  plotMyMatrix(dimLabels = c("cult2014", "soil_code"), plotOptions = list(rowNames = T,colNames = F))
cult2015_bis%>%
  t()%>%
  plotMyMatrix(dimLabels = c("cult2014", "soil_code"), plotOptions = list(rowNames = T,colNames = F))
cult2016_bis%>%
  t()%>%
  plotMyMatrix(dimLabels = c("cult2014", "soil_code"), plotOptions = list(rowNames = T,colNames = F))
```


```{r echo = T, results = 'hide'}
### create networks objects
net2014_bis <- defineSBM(cult2014_bis, model = "bernoulli", dimLabels = c(col = "soil_code", row = "cult_cult_history"))# row and col inverted, bug ? 
net2015_bis <- defineSBM(cult2015_bis, model = "bernoulli", dimLabels = c(col = "soil_code", row = "cult_cult_history"))
net2016_bis <- defineSBM(cult2016_bis, model = "bernoulli", dimLabels = c(col = "soil_code", row = "cult_cult_history"))

### compute smb
set.seed(2)
estimateMultiplexSBM(list(net2014_bis, net2015_bis, net2016_bis), dependent = FALSE) -> sbm_preced_cult_bis
```

### Plot cultural history arrange

```{r echo = T}
plot(sbm_preced_cult_bis)
```


```{r echo = T, results = 'hide'}
cluster_df%>%
  left_join( data.frame(clust_cult_history = sbm_preced_cult_bis$memberships$cult_cult_history,
                       soil_code = preced_cultu$soil_code ), by = "soil_code") -> cluster_df
```

## Soil

```{r echo = T}
soil%>%
  select(-c(soil_code, N_site, Site))%>%
  mutate_all(~zscore_normalization(.))%>%
  as.matrix()-> soil_standard_rank
```



```{r echo=TRUE}
set.seed(2)
smb_soil <- estimateBipartiteSBM(soil_standard_rank, model = "gaussian", estimOptions = list(plot = FALSE))


var_soil_id = make_id_order_by_group_df(smb_soil$memberships[2], col)


soil2_id =  make_id_order_by_group_df(smb_soil$memberships[1], row)
  
cluster_df%>%
  left_join( data.frame(clust_soil = smb_soil$memberships$row,
                       soil_code = soil$soil_code ), by = "soil_code") -> cluster_df
```

### Soil plot (before sbm)
```{r echo = T}
soil%>%
  select(-c(soil_code, N_site, Site))%>%
   decostand("standardize")-> soil_val

soil_val%>%
  as.matrix()%>%
  plotMyMatrix(dimLabels = c("soil_code", "Soil var"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))
```

### Plot soil arrange (after sbm)

```{r}
soil_standard_rank[soil2_id$id,var_soil_id$id]%>%
  as.matrix()%>%
  plotMyMatrix(dimLabels =  c("soil_code", "Soil var"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(var_soil_id$col))[-1]-0.5, col ="red")+
  geom_hline(yintercept = abs(which(!duplicated(soil2_id$row))[-1]-nrow(soil_id)-1)+0.5, col = "red", linetype = 5)+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))


soil_val[soil2_id$id,var_soil_id$id]%>%
  as.matrix()%>%
  plotMyMatrix(dimLabels =  c("soil_code", "Soil var"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(var_soil_id$col))[-1]-0.5, col ="red")+
  geom_hline(yintercept = abs(which(!duplicated(soil2_id$row))[-1]-nrow(soil_id)-1)+0.5, col = "red", linetype = 5)+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))

```


## Variety

```{r}
unique(variety$varnieb)
unique(variety$varsorgh)

variety%>%
  mutate(varnieb = case_when(varnieb == "beng-raaga" ~"Beng-raaga",
                   varnieb == "Beng-yaanga spécial" ~"Beng-yaanga",
                   varnieb == "beng-yaanga" ~"Beng-yaanga",
                   varnieb == "beng raaga(beng zaalé)" ~"Beng-raaga",
                   varnieb == "Beng raaga" ~"Beng-raaga",
                  .default = as.character(varnieb)),
         varsorgh = case_when(varsorgh == "Mélange(Pissopoé + Mitindaadé)" ~"Mélange(Pissopoé + Mitindaadé)",
                   varsorgh == "rouko" ~"Rouko",
                  .default = as.character(varsorgh))) -> variety

tab.disjonctif(variety$varnieb) -> varnieb
varnieb%>%
  as.data.frame()%>%
  mutate(`Beng-raaga` = `Beng-raaga` + `Mélange (beng-raaga+Beng-yaanga)`,
         `Beng-yaanga` = `Beng-yaanga` + `Mélange (beng-raaga+Beng-yaanga)`) %>%
  select(!(starts_with("Mélange") ))%>%
  as.matrix() -> varnieb

tab.disjonctif(variety$varsorgh) -> varsorgh

varsorgh%>%
  as.data.frame()%>%
  mutate(`ICSV-1049` = `ICSV-1049` + `Mélange (ICSV-1049+Mitindaadé)`,
         `Mitindaadé` = `Mitindaadé` + `Mélange (ICSV-1049+Mitindaadé)`,
         Rouko = Rouko + `Mélange (Mitindaadé+rouko)`,
         `Mitindaadé` = `Mitindaadé` + `Mélange (Mitindaadé+rouko)`,
         Pisnou = Pisnou + `Mélange (Pisnou+Kazinga)`,
         Kazinga = `Mélange (Pisnou+Kazinga)`,
         `Mitindaadé` = `Mitindaadé` + `Mélange (Pissopoé + Mitindaadé)`,
         `Pissopoé` = `Pissopoé` + `Mélange (Pissopoé + Mitindaadé)`,
                  `Mitindaadé` = `Mitindaadé` + `Mélange(Pissopoé + Mitindaadé)`,
         `Pissopoé` = `Pissopoé` + `Mélange(Pissopoé + Mitindaadé)`,
         `Zoé-wongo` =  `Mélange (Zoé-wongo+Kourbouli)`,
         Kourbouli =  Kourbouli + `Mélange (Zoé-wongo+Kourbouli)`,
        `Zoé-wongo` = `Zoé-wongo` + `Zoé-wongo+rouko`,
         Rouko = Rouko + `Zoé-wongo+rouko`)%>%
        select(!(starts_with("Mélange")  | "Zoé-wongo+rouko"))%>%
  as.matrix() -> varsorgh
varsorgh%>%
  as.matrix()%>%
  t()%>%
  plotMyMatrix(dimLabels = c("Var. sorgh", "soil_code" ), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))
colSums(varsorgh)
colSums(varnieb)
```




```{r,results='hide'}
### create networks objects
type = "bipartite"
model = "bernoulli"
netvarnieb<- defineSBM(varnieb, model, type, directed, dimLabels = c("soil_code",
                                                                                    "varnieb"))
netvarsorgh <- defineSBM(varsorgh, model, type, directed, dimLabels = c("soil_code",
                                                                              "varsorgh"))

### compute smb
estimateMultipartiteSBM(list(netvarnieb, netvarsorgh), estimOptions = list(initBM = FALSE)) -> sbm_variety
```

### Plot variety (before sbm)
```{r}
plotMyMultipartiteMatrix(list(netvarnieb, netvarsorgh), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))

```

### Plot variety arrange (after sbm)
```{r}
plot(sbm_variety)
```

```{r}
sbm_variety$storedModels %>%  ggplot() + aes(x = nbBlocks, y = ICL)  + geom_line() + geom_point(alpha = 0.5)
```


```{r}
cluster_df%>%
  left_join( data.frame(clust_variety = sbm_variety$memberships$soil_code,
                        soil_code = soil$soil_code ), by = "soil_code") -> cluster_df
```

```{r}
smb_varsorg_test <- estimateBipartiteSBM(varsorgh, model = "bernoulli",
                                         estimOptions = list(plot = FALSE))
plot(smb_varsorg_test)

```

```{r}
colSums(varsorgh)
colSums(varnieb)
```

```{r}
cbind(varnieb, varsorgh)%>%
  data.frame()%>%
  mutate(across(everything(), ~case_when(. == 1 ~  cur_column(),
                                          . == 0 ~ NA)))%>%
  tidyr::unite("cult_var", na.rm = T) -> cult_var_df
  
cult_var_df%>%
  group_by(cult_var)%>%
  summarise(n= n())-> temp

ggplot(temp)+
  geom_point(aes(cult_var, n), cex = 3)+
  main_theme+
  theme(axis.text.x = element_text(colour = "black", size=12, face="italic", angle = 45, vjust = 1, hjust = 1))

cult_var_df%>%
  left_join(temp, by = "cult_var")%>%
  mutate(cult_var = case_when(n<5 ~ "others",
                              cult_var == "Beng.raaga_Mitindaadé_Pissopoé" ~ "Beng.raaga_Pissopoé",
                              cult_var == "Beng.raaga_Balinpèlga" ~ "others",
                               .default = cult_var),
          soil_code = soil$soil_code)%>%
  select(-n) -> cult_var_df
cult_var_df%>%
  group_by(cult_var)%>%
  summarise(n= n())%>%
  ggplot()+
  geom_point(aes(cult_var, n), cex = 3)+
  scale_y_continuous(limits = c(0,30))+
  main_theme+
  theme(axis.text.x = element_text(colour = "black", size=12, face="italic", angle = 45, vjust = 1, hjust = 1))

cluster_df%>%
  left_join( cult_var_df, by = "soil_code") -> cluster_df
```

## Cultural practices

### SBM

```{r}
agri_practice%>%select(where(is.numeric)) -> agri_practice_quanti
agri_practice_quanti%>%
  ggpairs()
agri_practice_quanti%>%
  mutate_all(~zscore_normalization(.)) -> agri_practice_normrank
agri_practice_normrank%>%ggpairs

tab.disjonctif(agri_practice$pracul) -> pracul_dummy
tab.disjonctif(agri_practice$modsem) -> modsem_dummy

netpracul<- defineSBM(pracul_dummy, model, type, dimLabels = c("soil_code",
                                                                     "pracul"))
netmodsem <- defineSBM(modsem_dummy, model, type, dimLabels = c("soil_code",
                                                                        "modsem"))
netamendement <- defineSBM(as.matrix(agri_practice_normrank), "gaussian", type, dimLabels = c("soil_code",
                                                                "amendement"))
estimateBipartiteSBM(as.matrix(agri_practice_normrank), model = "gaussian",
                     estimOptions = list(plot = FALSE))-> sbm_practice
  
estimateMultipartiteSBM(c(netpracul, netmodsem, netamendement), estimOptions = list(initBM = FALSE)) -> sbm_practice
```

### Plot practice (before sbm)

```{r}
plotMyMultipartiteMatrix(list(netpracul, netmodsem, netamendement), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))
```

### Plot practice arrange (after sbm)
```{r}
plot(sbm_practice)
```
No cluster !

### MFA (Quali and quanti in the same mutivariate analisys)
```{r}
MFA_agri_practice = MFA(agri_practice_standa, group = c(2,5), c("n","s"),
    name.group=c("technical itin","chemical itin"))
summary(MFA_agri_practice)
HCPC_practice = MFA_agri_practice%>%
  HCPC(-1)

fviz_mfa_var(MFA_agri_practice, "group")
fviz_mfa_quali_biplot(MFA_agri_practice, repel = FALSE, col.var = "#E7B800",
                      habillage = HCPC_practice$data.clust$clust, addEllipses = TRUE, ellipse.level = 0.95)
```
clustering is not working very well with MFA on agricultural practice, lot of cluster overlap

### PCA (Qanti only)

```{r}
agri_practice_quanti%>%
  decostand("standardize")%>%
  PCA -> PCA_practices
 PCA_practices%>%
  HCPC(-1)-> HCPC_pca_practice
```


```{r}
fviz_pca_biplot(PCA_practices, habillage = HCPC_pca_practice$data.clust$clust, addEllipses  =T, ellipse.level = 0.95) 
```


```{r}
grid_id = make_id_order_by_group_df(data.frame(clust = HCPC_pca_practice$data.clust$clust), clust)
agri_practice_quanti%>%
   decostand("standardize") -> temp
temp[grid_id$id,]%>%
  t()%>%
  as.matrix()%>%
  plotMyMatrix(dimLabels = c("Soil", "soil_code"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(grid_id$clust))[-1]-0.5, col ="red")+
  #geom_hline(yintercept = abs(which(!duplicated(soil_id$row))[-1]-nrow(soil_id)-1)+0.5, col = "red", linetype = 5)+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))

```

best clustering for agricultural practice so far

```{r}
cluster_df%>%
  left_join( data.frame(clust_practice = HCPC_pca_practice$data.clust$clust,
                        soil_code = metadata$soil_code ), by = "soil_code") -> cluster_df
```


## Knonw bact community analyse

### visualization of cluster and analysis

```{r}
### exctract locality
cluster_df%>%
  mutate(locality = factor((str_extract(soil_code,"((?<=^.{3})[A-Z]+)"))))-> cluster_df

```


```{r}
B <-
  as.data.frame(
    table(
      
      cluster_df$soil_code,
      cluster_df$clust_cult_history ,
      cluster_df$clust_soil,
      cluster_df$clust_variety,
      cluster_df$Clust_bact_known_sbm,
      cluster_df$locality,
      cluster_df$cult_var)
  )

colnames(B) =  c( "soil_code","clust_cult_history", "clust_soil",  "clust_variety", "Clust_bact_known_sbm", "locality","cult_var","Freq")

w   <- which(B$Freq != 0)
B <- B[w, ]
```


```{r}
factor(B$clust_cult_history)
B$clust_cult_history <- factor(B$clust_cult_history, labels = c( "Cereales +\n legumes only", "mixte"))
alluvial(B[, c(5,2)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_bact_known_sbm, c2 = cluster_df$clust_cult_history)
```



```{r}

B$clust_soil <- factor(B$clust_soil, labels = c( "high cation \n concentration + \n fertile", 
                                                 "high cation \n concentration",
                                                 "mid cation/fertile",
                                                 "low cation/fertile"))
B$Clust_bact_known_sbm <- factor(B$Clust_bact_known_sbm, levels = c( "2", "3",  "1", "4"))
alluvial(B[, c(5,3)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_bact_known_sbm, c2 = cluster_df$clust_soil)
```


```{r}
B$clust_variety <- factor(B$clust_variety, labels = c( "Beng.raaga", 
                                                 "others"))
alluvial(B[, c(5,4)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_bact_known_sbm, c2 = cluster_df$clust_variety)
```

```{r}
alluvial(B[, c(5,7)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_bact_known_sbm, c2 = cluster_df$cult_var)
```

```{r}
alluvial(B[, c(5,6)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_bact_known_sbm, c2 = cluster_df$locality)
```

At community scale, only soil caracteristics and locality are congruente (correlated) to bacterial community

```{r}
cluster_df%>%
  mutate(across(where(is.numeric), as.factor)) -> cluster_df
Y = tab.disjonctif(cluster_df$Clust_bact_known_sbm)
cca(Y~locality*clust_soil, data = cluster_df) -> cca_bact_known
plot(cca_bact_known)
cluster_df$interaction_soil_loc = paste(cluster_df$locality, cluster_df$clust_soil)
cluster_df$interaction_soil_hist = paste(cluster_df$clust_cult_history, cluster_df$clust_soil)
mod_varpart_bact = varpart(Y,
                                  ~locality,
                                  ~clust_soil, data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))
plot(mod_varpart_bact,Xnames = c(list("locality"), list("clust_soil")),
     bg = c( "forestgreen","#CB6CE6"), lwd = 2, cex = 1.5, alpha = 170)

mod2_varpart_bact = varpart(Y,
                            ~clust_soil, 
                            ~locality,
                            ~interaction_soil_loc,
                            data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))
plot(mod2_varpart_bact,Xnames = c(list("locality"), list("clust_soil"), list("interaction_soil_loc")),
     bg = c( "forestgreen","#CB6CE6","#C0344A"), lwd = 2, cex = 1.5, alpha = 170)

mod4_varpart_bact = varpart(Y,
                                  ~clust_cult_history,
                                  ~clust_soil, 
                                  data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))

plot(mod4_varpart_bact,Xnames = c(list("Soil hist"), list("clust_soil")),
     bg = c( "orange","#CB6CE6"), lwd = 2, cex = 1.5, alpha = 170)

mod5_varpart_bact = varpart(Y,
                                ~clust_cult_history,
                                  ~clust_soil, 
                                  ~interaction_soil_hist, data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))
plot(mod5_varpart_bact,Xnames = c(list("Soil hist"), list("clust_soil"), list("interaction_soil_hist")),
     bg = c( "orange","#CB6CE6","#C0344A"), lwd = 2, cex = 1.5, alpha = 170)

```

### JSDM

```{r}
library(Hmsc)
# vignette("vignette_3_multivariate_high")
# vignette("vignette_1_univariate")
# vignette("vignette_2_multivariate_low")
```

#### Prepare data for JSDM

```{r}
traits = data.frame(SBM_bact_similarity = sbm_bact_known$memberships$Bact_Known,
                    sp = colnames(bact_OTUs_rarefied_t))%>%
 filter(SBM_bact_similarity != max(SBM_bact_similarity))
  
```
Trait value extract from sbm clustering (species part). Make the assuption that species how appear fequently in the same sample should have same tratis and similare responce to covariate. Altough last group contain moslty rare species

```{r}
 bact_OTUs_rarefied_t%>%
  select(traits$sp) -> bact_OTUs_rarefied_selected_t

traits%>%
  mutate(SBM_bact_similarity = as.factor(SBM_bact_similarity))%>%
  column_to_rownames("sp")-> traits
```

```{r}

studyDesign = data.frame(sample = rownames(bact_OTUs_rarefied_selected_t), stringsAsFactors=TRUE)
rL = HmscRandomLevel(units = studyDesign$sample)
rL$nfMax = 15
soil_val$pHeau
bact_OTUs_rarefied_selected_t%>%as.matrix%>%log1p%>%hist(breaks = 100)

rL = HmscRandomLevel(units = studyDesign$sample)
rL$nfMax = 15

m1 = Hmsc(Y = bact_OTUs_rarefied_selected_t%>%log1p, studyDesign = studyDesign,
         XData = soil%>%select(pHeau,CEC), XFormula =  ~ CEC,
         TrData = traits, TrFormula = ~SBM_bact_similarity, distr = "poisson",
         ranLevels = list(sample = rL))

nChains = 2
test.run = T
if (test.run){
#with this option, the vignette evaluates in ca. 10 minutes in a laptop
thin = 2
samples = 200
transient = 50
} else {
#with this option, the vignette evaluates in ca. 2 hrs in a laptop
thin = 10
samples = 1000
transient = 500
}
verbose = 10
m_fit1 = sampleMcmc(m1, thin = thin, samples = samples, transient = transient,
nChains = nChains, nParallel = nChains, verbose = verbose)


postGamma = getPostEstimate(m_fit1, parName = "Gamma")
plotGamma(m_fit1, post=postGamma, param="Support", supportLevel = 0.95)

postBeta = getPostEstimate(m_fit1, parName="Beta")
plotBeta(m_fit1, post=postBeta, param="Support", supportLevel = 0.95)

Gradient = constructGradient(m_fit1, focalVariable = "CEC", ngrid = 80)
predY = predict(m_fit1, XData=Gradient$XDataNew, expected=TRUE)
par(mfrow = c(1,1))
plotGradient(m_fit1, Gradient, pred=predY, measure="S", showData = TRUE)
plotGradient(m_fit1, Gradient, pred=predY, measure="Y", index = 4, showData = TRUE)
```

# Second approach : work on family
```{r echo = T, results = 'hide'}

bact_tax%>%
  select(-c(genus, species))%>%
  distinct(family)%>%
  filter(!str_detect(family, "unidentified"))%>%
  arrange(family)-> bact_tax_family

bact_ASV_decont%>%
  filter(as.character(family ) %in% bact_tax_family$family)%>%
  pivot_longer(starts_with("BU"), names_to = "soil_code", values_to = "read")%>%
  group_by(family, soil_code)%>%
  summarise(read = sum(read))%>%
  pivot_wider(names_from = "soil_code", values_from = "read") %>% 
  column_to_rownames("family")-> OTU_bact_family

```

# Rarefy
```{r}
set.seed(2)

OTU_bact_family_rarefied_t <-  OTU_bact_family %>% 
  t() %>% 
  rrarefy (min(colSums(OTU_bact_family))) %>% 
  #{{create a new column with sum of reads for each OTU in all samples}}

  as_tibble(rownames = NA) %>%
  rownames_to_column(var = "OTU") %>%
  #{{delete all sample with = 0 in rarefied samples}}
  mutate(total = rowSums(across(where(is.numeric)))) %>% 
  filter(total != 0) %>%
  select(-total) %>%
  column_to_rownames(var = "OTU") %>% 
  as_tibble(rownames = NA)

hist(OTU_bact_family_rarefied_t%>%as.matrix%>%log1p, breaks = 100)
```


### Clustering bact known

#### Unarrange matrix
```{r}

OTU_bact_family_rarefied_t%>%
  as.matrix()%>%
  log1p%>%
  plotMyMatrix(dimLabels = c("soil_code", "Bact family"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))
```

#### arrange matrix

```{r, results='hide'}
bact_OTUs_rarefied_t

if(file.exists("outputs/sbm_bact_family.Rdata")){
  load(file = "outputs/sbm_bact_family.Rdata")
}else{
  OTU_bact_family_rarefied_t%>%
  as.matrix%>%
  log1p%>%
  estimateBipartiteSBM(
    model = 'poisson', 
    dimLabels = c("soil_code", "bact_family"),
    estimOptions = list(nbCores = 8,plot = F)) -> sbm_bact_family
  
  save(sbm_bact_family, file = "outputs/sbm_bact_family.Rdata")
  
}


sbm_bact_known$storedModels %>% knitr::kable()
cluster_df%>%
  left_join(data.frame(clust_bact_family_sbm = sbm_bact_family$memberships$soil_code, 
                         soil_code = rownames(OTU_bact_family_rarefied_t)), by = "soil_code") -> cluster_df
```

```{r}

bact_id_fam = make_id_order_by_group_df(sbm_bact_family$memberships[2], bact_family)


soil_id_fam =  make_id_order_by_group_df(sbm_bact_family$memberships[1], soil_code)

OTU_bact_family_rarefied_t[soil_id_fam$id,bact_id_fam$id]%>%
  as.matrix()%>%
    log1p%>%
  plotMyMatrix(dimLabels = c("soil_code", "Bact known"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(bact_id_fam$bact_family))[-1]-0.5, col ="red", linetype = 5)+
  geom_hline(yintercept = abs(which(!duplicated(soil_id_fam$soil_code))[-1]-nrow(soil_id_fam)-1)+0.5, col = "red")+
  annotate("text",x = c(1,which(!duplicated(bact_id_fam$bact_family))[-1])+1, y =5, label = unique(bact_id_fam$bact_family), col= "red")+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))


```

### visualization of cluster and analysis

```{r}

B2 <-
  as.data.frame(
    table(
      
      cluster_df$soil_code,
      cluster_df$clust_cult_history ,
      cluster_df$clust_soil,
      cluster_df$clust_variety,
      cluster_df$clust_bact_family_sbm,
      cluster_df$locality,
      cluster_df$cult_var)
  )

colnames(B2) =  c( "soil_code","clust_cult_history", "clust_soil",  "clust_variety", "clust_bact_family_sbm", "locality","cult_var","Freq")

w   <- which(B2$Freq != 0)
B2 <- B2[w, ]
```


```{r}

B2$clust_cult_history <- factor(B2$clust_cult_history, labels = c( "Cereales +\n legumes only", "mixte"))
alluvial(B2[, c(5,2)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$clust_bact_family_sbm, c2 = cluster_df$clust_cult_history)
```



```{r}

B2$clust_soil <- factor(B2$clust_soil, labels = c( "high cation \n concentration + \n fertile", 
                                                 "high cation \n concentration",
                                                 "mid cation/fertile",
                                                 "low cation/fertile"))

alluvial(B2[, c(5,3)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$clust_bact_family_sbm, c2 = cluster_df$clust_soil)
```


```{r}
B2$clust_variety <- factor(B2$clust_variety, labels = c( "Beng.raaga", 
                                                 "others"))
alluvial(B2[, c(5,4)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$clust_bact_family_sbm, c2 = cluster_df$clust_variety)
```

```{r}
alluvial(B2[, c(5,7)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$clust_bact_family_sbm, c2 = cluster_df$cult_var)
```

```{r}
alluvial(B2[, c(5,6)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$clust_bact_family_sbm, c2 = cluster_df$locality)
```

At community scale, only soil caracteristics and locality are congruente (correlated) to bacterial community

```{r}
cluster_df%>%
  mutate(across(where(is.numeric), as.factor)) -> cluster_df
Y_family = tab.disjonctif(cluster_df$clust_bact_family_sbm)
cca(Y_family~locality*clust_soil, data = cluster_df) -> cca_bact_known
plot(cca_bact_known)
cluster_df$interaction_soil_loc = paste(cluster_df$locality, cluster_df$clust_soil)
cluster_df$interaction_soil_hist = paste(cluster_df$clust_cult_history, cluster_df$clust_soil)
mod_varpart_bact = varpart(Y_family,
                                  ~locality,
                                  ~clust_soil, data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))
plot(mod_varpart_bact,Xnames = c(list("locality"), list("clust_soil")),
     bg = c( "forestgreen","#CB6CE6"), lwd = 2, cex = 1.5, alpha = 170)

mod2_varpart_bact = varpart(Y_family,
                            ~clust_soil, 
                            ~locality,
                            ~interaction_soil_loc,
                            data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))
plot(mod2_varpart_bact,Xnames = c(list("locality"), list("clust_soil"), list("interaction_soil_loc")),
     bg = c( "forestgreen","#CB6CE6","#C0344A"), lwd = 2, cex = 1.5, alpha = 170)

mod4_varpart_bact = varpart(Y_family,
                                  ~clust_cult_history,
                                  ~clust_soil, 
                                  data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))

plot(mod4_varpart_bact,Xnames = c(list("Soil hist"), list("clust_soil")),
     bg = c( "orange","#CB6CE6"), lwd = 2, cex = 1.5, alpha = 170)

mod5_varpart_bact = varpart(Y_family,
                                ~clust_cult_history,
                                  ~clust_soil, 
                                  ~interaction_soil_hist, data= cluster_df, chisquare=T,
                                  permutations = how(nperm = 10000))
plot(mod5_varpart_bact,Xnames = c(list("Soil hist"), list("clust_soil"), list("interaction_soil_hist")),
     bg = c( "orange","#CB6CE6","#C0344A"), lwd = 2, cex = 1.5, alpha = 170)

```

###JSDM
#### Prepare data for JSDM

```{r}
traits_fam = data.frame(SBM_family_similarity = sbm_bact_family$memberships$bact_family,
                    family = colnames(OTU_bact_family_rarefied_t))%>%
 filter(SBM_family_similarity != max(SBM_family_similarity))
  
```
Trait value extract from sbm clustering (species part). Make the assuption that species how appear fequently in the same sample should have same tratis and similare responce to covariate. Altough in last group containe moslty rare species

```{r}
 OTU_bact_family_rarefied_t%>%
  select(traits_fam$family) -> OTU_bact_family_rarefied_selected_t
traits_fam%>%
  mutate(SBM_family_similarity = as.factor(SBM_family_similarity))%>%
  column_to_rownames("family")-> traits_fam
```

```{r}

studyDesign = data.frame(sample = rownames(OTU_bact_family_rarefied_selected_t), stringsAsFactors=TRUE)
rL = HmscRandomLevel(units = studyDesign$sample)
rL$nfMax = 15
soil%>%
  select(-where(is.character),-Pto)%>%
  cor()%>%
  corrplot(method = "square", order = "hclust")
Xdata = soil%>%
  select(CN,TS, K,pHeau,CEC, MO,soil_code)%>%
  left_join(cluster_df%>%select(clust_cult_history,soil_code), by = "soil_code")%>%
  select(-"soil_code")
OTU_bact_family_rarefied_selected_t%>%as.matrix%>%log1p%>%hist(breaks = 100)

rL = HmscRandomLevel(units = studyDesign$sample)
rL$nfMax = 15

m = Hmsc(Y = OTU_bact_family_rarefied_selected_t%>%log1p,
         XData = Xdata, XFormula =  ~CN +TS+ K+pHeau+CEC+ MO+clust_cult_history,
         TrData = traits_fam, TrFormula = ~SBM_family_similarity, distr = "poisson")

nChains = 2
test.run = T
if (test.run){
#with this option, the vignette evaluates in ca. 10 minutes in a laptop
thin = 2
samples = 200
transient = 50
} else {
#with this option, the vignette evaluates in ca. 2 hrs in a laptop
thin = 10
samples = 1000
transient = 100
}
verbose = 10
m_fit = sampleMcmc(m, thin = thin, samples = samples, transient = transient,
nChains = nChains, nParallel = nChains, verbose = verbose)

postGamma = getPostEstimate(m_fit, parName = "Gamma")
plotGamma(m_fit, post=postGamma, param="Support", supportLevel = 0.95)


postBeta = getPostEstimate(m_fit, parName="Beta")
plotBeta(m_fit, post=postBeta, param="Sign", supportLevel = 0.95)

postBeta$mean%>%
  as.data.frame%>%
  mutate(type =c("intercept", paste(rep("beta",nrow(.)-1),2:nrow(.), sep="")))%>%
  column_to_rownames("type")%>%
  t%>%as.data.frame%>%
  rownames_to_column("family")%>%
  left_join(traits_fam%>%rownames_to_column("family"), by = "family")%>%
  pivot_longer(starts_with("beta"), names_to = "parameters", values_to = "parameter_estimate")%>%
  ggplot()+
  geom_point(aes(intercept, parameter_estimate, col = SBM_family_similarity))+
  facet_wrap(~parameters, scales = "free")+
  main_theme
  
Gradient = constructGradient(m_fit, focalVariable = "pHeau", ngrid = 80)
predY = predict(m_fit, XData=Gradient$XDataNew, expected=T)
par(mfrow = c(2,2))
plotGradient(m_fit, Gradient, pred=predY, measure="S", showData = TRUE)
plotGradient(m_fit, Gradient, pred=predY, measure="Y", index = 14, showData = TRUE)
plotGradient(m_fit, Gradient, pred=predY, measure="Y", index = 2, showData = TRUE)
plotGradient(m_fit, Gradient, pred=predY, measure="Y", index = 21, showData = TRUE)

GradientCEC = constructGradient(m_fit, focalVariable = "CEC", ngrid = 80)
predY2 = predict(m_fit, XData=GradientCEC$XDataNew, expected=T)
par(mfrow = c(2,2))
plotGradient(m_fit, GradientCEC, pred=predY2, measure="S", showData = TRUE)
plotGradient(m_fit, GradientCEC, pred=predY2, measure="Y", index = 14, showData = TRUE)
plotGradient(m_fit, GradientCEC, pred=predY2, measure="Y", index = 2, showData = TRUE)
plotGradient(m_fit, GradientCEC, pred=predY2, measure="Y", index = 11, showData = TRUE)

Gradienthist = constructGradient(m_fit, focalVariable = "clust_cult_history", ngrid = 80)
predY3 = predict(m_fit, XData=Gradienthist$XDataNew, expected=T)
par(mfrow = c(2,2))
plotGradient(m_fit, Gradienthist, pred=predY3, measure="S", showData = TRUE,pointsize= 3)
plotGradient(m_fit, Gradienthist, pred=predY3, measure="Y", index = 14, showData = TRUE)
plotGradient(m_fit, Gradienthist, pred=predY3, measure="Y", index = 2, showData = TRUE)
plotGradient(m_fit, Gradienthist, pred=predY3, measure="Y", index = 11, showData = TRUE)
```

## Diversities

```{r}

```

# FUNGI

## setup fungi table
```{r echo = T, results = 'hide'}
its_taxonomic_levels <- c("kingdom", "phylum", "class", "order", "family", 
                           "genus", "species")

 rename_with(~sub("\\_.*", "", .)) %>%
  rename_with(~gsub("-", "_", .)) %>%
  rename_with(~gsub("OR", "BU",.)) %>%
  rename(CONT_BU_ext = BU_T_extr) %>%
  rename(CONT_BU_PCR1 = T_1) %>%
  rename(CONT_BU_PCR2 = T_2) %>%
  #{{select samples from ORACLE project}}
  select(OTU, taxonomy, starts_with("CONT", ignore.case = FALSE), 
         starts_with("BU", ignore.case = FALSE)) %>%
  #{{keep only resequenced samples when available}}
  select(-BU_SER_10, -BU_ST_10, -BU_YL_19, -BU_TS_3) 
  #{{homogeneize resequenced sample names}}
   
#{{import data}}
read_tsv("data/Oracle_run_20200106_ITS2_ITS86F_ITS4_169_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2") %>%
  #{{format column names}}
  rename_with(~sub("\\_.*", "", .)) %>%
  rename_with(~gsub("-", "_", .)) %>%
  rename_with(~gsub("OR", "BU",.)) %>%
  rename(CONT_BU_ext = BU_T_extr) %>%
  rename(CONT_BU_PCR1 = T_1) %>%
  rename(CONT_BU_PCR2a = T_2) %>%
  rename(CONT_BU_PCR2b = T_3) %>%
  #{{select samples from ORACLE project}}
  select(OTU, taxonomy, starts_with("CONT", ignore.case = FALSE),
         starts_with("BU_RT_", ignore.case = FALSE),
         starts_with("BU_SER_", ignore.case = FALSE),
         starts_with("BU_ST_", ignore.case = FALSE),
         starts_with("BU_TS_", ignore.case = FALSE),
         starts_with("BU_YL_", ignore.case = FALSE)) %>%
  #{{homogeneize resequenced sample names}}
  rename_with(~sub("\\_1$", "_01", .)) %>%
  rename_with(~sub("\\_2$", "_02", .)) %>%
  rename_with(~sub("\\_3$", "_03", .)) %>%
  rename_with(~sub("\\_4$", "_04", .)) %>%
  rename_with(~sub("\\_5$", "_05", .)) %>%
  rename_with(~sub("\\_6$", "_06", .)) %>%
  rename_with(~sub("\\_7$", "_07", .)) %>%
  rename_with(~sub("\\_8$", "_08", .)) %>%
  rename_with(~sub("\\_9$", "_09", .)) %>%
  #{{"Fungi" are selected}}
  filter(grepl("Fungi", taxonomy)) %>%
  #{{select only OTUs and controls}}
  select(OTU, starts_with("CONT", ignore.case = FALSE),
         starts_with("BU", ignore.case = FALSE)) %>%
  droplevels() -> ITS2_OTUs



ITS2_OTUs %>% 
  #{{merge taxonomy}}
  left_join(select(read_tsv("data/Oracle_run_20200106_ITS2_ITS86F_ITS4_169_samples.OTU.filtered.cleaved.nosubstringOTUs.mumu.table2"), OTU, taxonomy), by = "OTU") %>%
  #{{select only OTU number and taxonomy}}
  select(OTU, taxonomy) %>%
  #{{"Fungi" are selected}}
  separate(taxonomy, its_taxonomic_levels, sep = "[|]", extra = "drop") %>%
  #{{remove "k__" bits, etc in taxonomic names and replace missing taxonomic information}}
  mutate_if(is.character, ~sub("^[kpcofgs]__", "", .)) %>%
  mutate_if(is.character, ~sub("[*]", "unidentified", .)) %>%
  #{{replace unidentified by the higher taxonomic rank}}
  mutate(phylum = if_else(str_detect(phylum, "unidentified"),
                          str_c("unidentified_", kingdom), phylum)) %>%
  mutate(class = if_else(str_detect(class, "unidentified"),
                         str_c("unidentified_",phylum), class)) %>%
  mutate(order = if_else(str_detect(order, "unidentified"),
                         str_c("unidentified_",class), order)) %>%
  mutate(family = if_else(str_detect(family, "unidentified"),
                          str_c("unidentified_", order), family)) %>%
  mutate(genus = if_else(str_detect(genus, "unidentified"),
                         str_c("unidentified_", family), genus)) %>%
  mutate(species = if_else(str_detect(species, "unidentified"),
                           str_c("unidentified_",genus), species)) %>%
  #{{homogneize unidentified terms for taxonomic rank}}
  mutate(class = str_replace(class,
                             "unidentified_unidentified_",
                             "unidentified_")) %>%
  mutate(order = str_replace(order,
                             "unidentified_unidentified_unidentified_",
                             "unidentified_")) %>%
  mutate(order = str_replace(order,
                             "unidentified_unidentified_",
                             "unidentified_")) %>%
  mutate(family = str_replace(family,
                              "unidentified_unidentified_unidentified_unidentified_",
                              "unidentified_")) %>% 
  mutate(family = str_replace(family,
                              "unidentified_unidentified_unidentified_",
                              "unidentified_")) %>% 
  mutate(family = str_replace(family,
                              "unidentified_unidentified_",
                              "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_unidentified_unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(genus = str_replace(genus,
                             "unidentified_unidentified_",
                             "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_unidentified_unidentified_unidentified_",
                               "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_unidentified_unidentified_",
                               "unidentified_")) %>% 
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_unidentified_",
                               "unidentified_")) %>%
  mutate(species = str_replace(species,
                               "unidentified_unidentified_unidentified_",
                               "unidentified_")) %>%
  mutate(species = str_replace(species,
                               "unidentified_unidentified_",
                               "unidentified_")) %>% 
  #{{homogenize terms for unidentified or uncharacterized species (sp)}}
  mutate(species = if_else(str_detect(species, "_sp"), 
                           str_c("unidentified_", genus), species)) %>% 
  mutate(species = str_replace(species, "unidentified_unidentified_",
                               "unidentified_")) %>%
  #{{convert OTU column in character from numeric}}
  mutate_if(is.numeric,as.character)-> ITS2_tax

ITS2_OTUs%>%
  mutate(OTU = as.character(OTU))%>%
  left_join(ITS2_tax, by = "OTU") -> ITS2_OTUs
  
d <- replace(ITS2_OTUs, ITS2_OTUs == 0, NA) %>%
  select(OTU, starts_with("CONT")) %>%
  gather("samples", "reads", -OTU) %>%
  filter(!is.na(reads))
ITS2_OTUs%>%
  select(starts_with("BU"))
ITS2_OTUs_decont <- replace(ITS2_OTUs, ITS2_OTUs == 0, NA) %>%
  pivot_longer(starts_with("BU"), names_to = "samples", values_to = "reads") %>%
  filter(!is.na(reads)) %>%
  #{{merge with control samples}}
  left_join(count(d, OTU, wt = reads), by = "OTU") %>%
  #{{substract abundance of control samples}}
  mutate(reads = case_when(
    is.na(n)  ~ reads,
    n > reads ~ 0,
    TRUE      ~ reads - n)) %>%
  select(-n) %>%
  spread(samples, reads, fill = 0) %>%
  select(-starts_with("CONT"))

```

```{r}
ITS2_tax%>%
  filter(!str_detect(species, "unidentified"))-> its_tax_known

its_tax_known[duplicated(its_tax_known$species),]
ITS2_OTUs_decont%>%
  filter(as.character(OTU) %in% its_tax_known$OTU)-> fungi_its_decont_known



fungi_its_decont_known%>%pivot_longer(starts_with("BU"), names_to = "soil_code", values_to = "read")%>%
  group_by(species, soil_code)%>%
  summarise(read = sum(read))%>%
  pivot_wider(names_from = "soil_code", values_from = "read") -> ITS_fungi_known

```

```{r echo = T}

ITS_fungi_known_t <- ITS_fungi_known%>%as.data.frame%>%
  column_to_rownames(var = "species") %>%
  #{{transpose data}}
  t()

plot(1:length(rowSums(ITS_fungi_known_t)), log10(sort(rowSums(ITS_fungi_known_t), decreasing = T)))


set.seed(2)
fungi_ITS_rarefied_t <-  ITS_fungi_known_t %>% 
  rrarefy (round(10^3.5))%>% 
  #{{create a new column with sum of reads for each OTU in all samples}}
  t() %>% 
  as_tibble(rownames = NA) %>%
  rownames_to_column(var = "OTU") %>%
  #{{delete all sample with = 0 in rarefied samples}}
  mutate(total = rowSums(across(where(is.numeric)))) %>% 
  filter(total != 0) %>%
  select(-total) %>%
  column_to_rownames(var = "OTU") %>% 
  t() %>%
  as_tibble(rownames = NA)

hist(fungi_ITS_rarefied_t%>%as.matrix%>%log1p)
```


### Clustering fungi known

#### Unarrange matrix
```{r}

fungi_ITS_rarefied_t%>%
  as.matrix()%>%
  log1p%>%
  plotMyMatrix(dimLabels = c("soil_code", "Fungi Known"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))
```

#### arrange matrix

```{r, results='hide'}


if(file.exists("outputs/sbm_fungi_known.Rdata")){
  load(file = "outputs/sbm_fungi_known.Rdata")
}else{
  fungi_ITS_rarefied_t%>%
  as.matrix%>%
  log1p%>%
  estimateBipartiteSBM(
    model = 'poisson', 
    dimLabels = c("soil_code", "fungi_known"),
    estimOptions = list(nbCores = 8,plot = F)) -> sbm_fungi_known
  
  save(sbm_fungi_known, file = "outputs/sbm_fungi_known.Rdata")
  
}

cluster_df$soil_code
sbm_fungi_known$storedModels %>% knitr::kable()
cluster_df$Clust_fungi_known_sbm = NULL
cluster_df%>%
  left_join(data.frame(Clust_fungi_known_sbm = sbm_fungi_known$memberships$soil_code, 
                         soil_code = rownames(fungi_ITS_rarefied_t)), by =  "soil_code") -> cluster_df
```

```{r}
sbm_fungi_known$storedModels %>%  ggplot() + aes(x = nbBlocks, y = ICL)  + geom_line() + geom_point(alpha = 0.5)
```

```{r}

fungi_id = make_id_order_by_group_df(sbm_fungi_known$memberships[2], fungi_known)


soil3_id =  make_id_order_by_group_df(sbm_fungi_known$memberships[1], soil_code)

fungi_ITS_rarefied_t[soil3_id$id,fungi_id$id]%>%
  as.matrix()%>%
    log1p%>%
  plotMyMatrix(dimLabels = c("soil_code", "fungi known"), plotOptions = list(rowNames = T,colNames = T))+
  theme(
    legend.position="none",
    axis.text.x =element_text(angle = 45, vjust = 1, hjust = 1, size = 6, face="italic"),
    axis.text.y = element_text(size = 5,face="italic"))+
  geom_vline(xintercept = which(!duplicated(fungi_id$fungi_known))[-1]-0.5, col ="red", linetype = 5)+
  geom_hline(yintercept = abs(which(!duplicated(soil3_id$soil_code))[-1]-nrow(soil3_id)-1)+0.5, col = "red")+
  theme(strip.text.x = element_text(colour = "gray", size=15, face ="italic"),
        strip.text.y  = element_text(colour = "gray", size=15, face ="italic"),
        strip.background = element_rect(fill="white"),
        panel.border = element_rect(color = "black",  fill = NA))


```
```{r}
plot(fungi_ITS_rarefied_t, type = "expected")
```

```{r}

B3 <-
  as.data.frame(
    table(
      
      cluster_df$soil_code,
      cluster_df$clust_cult_history ,
      cluster_df$clust_soil,
      cluster_df$clust_variety,
      cluster_df$clust_bact_family_sbm,
      cluster_df$locality,
      cluster_df$cult_var,
      cluster_df$Clust_fungi_known_sbm)
  )

colnames(B3) =  c( "soil_code","clust_cult_history", "clust_soil",  "clust_variety", "clust_bact_family_sbm", "locality","cult_var","Clust_fungi_known_sbm","Freq")

w   <- which(B3$Freq != 0)
B3 <- B3[w, ]
```


```{r}

B3$clust_cult_history <- factor(B3$clust_cult_history, labels = c( "Cereales +\n legumes only", "mixte"))
alluvial(B3[, c(8,2)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$clust_bact_family_sbm, c2 = cluster_df$clust_cult_history)
```



```{r}

B3$clust_soil <- factor(B3$clust_soil, labels = c( "high cation \n concentration + \n fertile", 
                                                 "high cation \n concentration",
                                                 "mid cation/fertile",
                                                 "low cation/fertile"))

alluvial(B3[, c(8,3)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_fungi_known_sbm, c2 = cluster_df$clust_soil)
```


```{r}
alluvial(B3[, c(8,7)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_fungi_known_sbm, c2 = cluster_df$cult_var)
```

```{r}
alluvial(B3[, c(8,6)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_fungi_known_sbm, c2 = cluster_df$locality)
```

```{r}
alluvial(B3[, c(8,5)],
         freq = B$Freq,
         alpha = 0.7,
         cex = 1.5,
         cex.axis = 1.5,
         xw = 0.2)
NMI(c1 = cluster_df$Clust_fungi_known_sbm, c2 = cluster_df$Clust_bact_known_sbm)
```

